### PR TITLE
[DTM0] EOS-19130: Fix panic during disconnect.

### PR DESCRIPTION
**Pull request by ivan-alekhin**
The patch fixes the problem where m0d services
were failing because DTM0-related connections
were not properly finalized.

### CODE REVIEW

File: dtm0/service.c [hunk 1]
```diff
@@ -62,9 +62,10 @@ static const struct m0_reqh_service_type_ops dtm0_service_type_ops = {
 };
 
 static const struct m0_reqh_service_ops dtm0_service_ops = {
-	.rso_start = dtm0_service_start,
-	.rso_stop  = dtm0_service_stop,
-	.rso_fini  = dtm0_service_fini
+	.rso_start           = dtm0_service_start,
+	.rso_stop            = dtm0_service_stop,
+	.rso_fini            = dtm0_service_fini,
+	.rso_prepare_to_stop = dtm0_service__ha_unsubscribe,
```
Comment (sergey-shilov) [inline dtm0/service.c:68]: Propose to name consistently, dtm0_service_prepare_to_stop() which calls dtm0_service__ha_unsubscribe(). Possibly dtm0_service_prepare_to_stop() will be extended.

File: dtm0/service.c [hunk 2]
```diff
@@ -479,7 +497,7 @@ static bool process_clink_cb(struct m0_clink *clink)
 	 * trigger for handling HA messages in this case.
 	 */
 	if (evented_proc_state != M0_NC_TRANSIENT &&
-	    !ready_to_process_ha_msgs && !m0_dtm0_in_ut()) {
+	    !ready_to_process_ha_msgs) {
```
Comment (sergey-shilov) [inline dtm0/service.c:500]: Why m0_dtm0_in_ut() check is not needed anymore?

File: dtm0/service.c (hunk 1 above)
Comment (ivan-alekhin) [reply dtm0/service.c:68]: makes sense, I'll change that.

File: dtm0/service.c (hunk 2 above)
Comment (ivan-alekhin) [reply dtm0/service.c:500]: It has been checked already at the start:

```c
	if (m0_dtm0_in_ut()) {
		M0_LOG(M0_DEBUG, "No HA handling in UT.");
		goto out;
	}
```

I don't think it is necessary to check this again. That's why I have removed it.

Comment (sergey-shilov) [reply dtm0/service.c:500]: Agreed.

File: dtm0/service.c [hunk 3]
```diff
@@ -233,33 +234,50 @@ m0_dtm0_service_process_connect(struct m0_reqh_service *s,
...
+	M0_SET0(&process->dop_rlink);
 
 	return M0_RC(rc);
 }
 
+M0_INTERNAL int
+m0_dtm0_service_process_disconnect(struct m0_reqh_service *s,
+				   struct m0_fid          *remote_srv)
+{
+	struct dtm0_process *process;
```
Comment (just-now) [inline dtm0/service.c:270]: nit: `struct dtm0_process *process = dtm0_service_process__lookup(s, remote_srv);`

**Review by just-now — APPROVED** [review]

File: dtm0/service.c (hunk 3 above)
Comment (ivan-alekhin) [reply dtm0/service.c]: done, and I also folded the return statements.

**Review by sergey-shilov — APPROVED** [review]

**Pull request merged**